diff -Naurp ruby-2.0.0-p648.orig/ext/openssl/ossl.c ruby-2.0.0-p648/ext/openssl/ossl.c
--- ruby-2.0.0-p648.orig/ext/openssl/ossl.c	2022-08-16 11:02:18.761934025 -0700
+++ ruby-2.0.0-p648/ext/openssl/ossl.c	2022-08-16 11:26:30.114113774 -0700
@@ -414,18 +414,8 @@ ossl_debug_get(VALUE self)
 static VALUE
 ossl_debug_set(VALUE self, VALUE val)
 {
-    VALUE old = dOSSL;
-    dOSSL = val;
+    dOSSL = RTEST(val) ? Qtrue : Qfalse;
 
-    if (old != dOSSL) {
-	if (dOSSL == Qtrue) {
-	    CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);
-	    fprintf(stderr, "OSSL_DEBUG: IS NOW ON!\n");
-	} else if (old == Qtrue) {
-	    CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_OFF);
-	    fprintf(stderr, "OSSL_DEBUG: IS NOW OFF!\n");
-	}
-    }
     return val;
 }
 
diff -Naurp ruby-2.0.0-p648.orig/ext/openssl/ossl_pkey_rsa.c ruby-2.0.0-p648/ext/openssl/ossl_pkey_rsa.c
--- ruby-2.0.0-p648.orig/ext/openssl/ossl_pkey_rsa.c	2022-08-16 11:02:18.773934354 -0700
+++ ruby-2.0.0-p648/ext/openssl/ossl_pkey_rsa.c	2022-08-16 11:02:10.569708558 -0700
@@ -745,7 +745,6 @@ Init_ossl_rsa()
     rb_define_method(cRSA, "params", ossl_rsa_get_params, 0);
 
     DefRSAConst(PKCS1_PADDING);
-    DefRSAConst(SSLV23_PADDING);
     DefRSAConst(NO_PADDING);
     DefRSAConst(PKCS1_OAEP_PADDING);
 
