diff --git a/ext/openssl/extconf.rb b/ext/openssl/extconf.rb
index 920caa6..1f94d97 100644
--- a/ext/openssl/extconf.rb
+++ b/ext/openssl/extconf.rb
@@ -43,14 +43,11 @@ if $mingw
   have_library("wsock32")
   have_library("gdi32")
 end
-
-result = pkg_config("openssl") && have_header("openssl/ssl.h")
-
-unless result
-  result = have_header("openssl/ssl.h")
-  result &&= %w[crypto libeay32].any? {|lib| have_library(lib, "OpenSSL_add_all_digests")}
-  result &&= %w[ssl ssleay32].any? {|lib| have_library(lib, "SSL_library_init")}
-  unless result
+result = have_header("openssl/ssl.h")
+result &&= %w[crypto libeay32].any? {|lib| have_library(lib, "OpenSSL_add_all_digests")}
+result &&= %w[ssl ssleay32].any? {|lib| have_library(lib, "SSL_library_init")}
+if !result
+  unless pkg_config("openssl") and have_header("openssl/ssl.h")
     message "=== Checking for required stuff failed. ===\n"
     message "Makefile wasn't created. Fix the errors above.\n"
     exit 1
