diff --git a/gc.c b/gc.c
index 9b5d588..e1144ff 100644
--- a/gc.c
+++ b/gc.c
@@ -1031,6 +1031,7 @@ ruby_xfree(void *x)
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_enable(void)
 {
@@ -1053,6 +1054,7 @@ rb_gc_enable(void)
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_disable(void)
 {
@@ -1075,6 +1077,7 @@ rb_gc_disable(void)
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_enable_stats()
 {
@@ -1096,6 +1099,7 @@ rb_gc_enable_stats()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_disable_stats()
 {
@@ -1115,6 +1119,7 @@ rb_gc_disable_stats()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_stats_enabled()
 {
@@ -1134,6 +1139,7 @@ rb_gc_stats_enabled()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_clear_stats()
 {
@@ -1204,6 +1210,7 @@ rb_gc_num_allocations()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_enable_trace()
 {
@@ -1225,6 +1232,7 @@ rb_gc_enable_trace()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_disable_trace()
 {
@@ -1244,6 +1252,7 @@ rb_gc_disable_trace()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_trace_enabled()
 {
@@ -1272,6 +1281,7 @@ const char* GC_LOGFILE_IVAR = "@gc_logfile_name";
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_log_file(int argc, VALUE *argv, VALUE self)
 {
