diff --git a/ext/fiddle/extconf.rb b/ext/fiddle/extconf.rb
index 03b0ac2..4b8c4cf 100644
--- a/ext/fiddle/extconf.rb
+++ b/ext/fiddle/extconf.rb
@@ -4,11 +4,16 @@ require 'mkmf'
 
 dir_config 'libffi'
 
-pkg_config("libffi")
-unless have_header('ffi.h')
+def have_ffi_headers
+  return true if have_header('ffi.h')
   if have_header('ffi/ffi.h')
     $defs.push(format('-DUSE_HEADER_HACKS'))
-  else
+  end
+end
+
+unless have_ffi_headers
+  pkg_config('libffi')
+  unless have_ffi_headers
     abort "ffi.h is missing. Please install libffi."
   end
 end
