From bb2f37b12b4acfa29171ea234e8cee1be4e12dff Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20Matyas?= <github@higher.lv>
Date: Sun, 26 Apr 2015 20:33:10 +0200
Subject: [PATCH] Clean bundler environment before running Homebrew during
 configuration

---
 configure | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/configure b/configure
index 3a35743..c487a3b 100755
--- a/configure
+++ b/configure
@@ -576,7 +576,7 @@ class Configure
         if macports?
           config = macports_llvm_config
         else
-          out = `brew list llvm | grep '/llvm-config$'`
+          out = Bundler.with_clean_env { `brew list llvm | grep '/llvm-config$'` }
           config = out.chomp if $?.success?
         end
       end
-- 
2.3.5

