--- a/tool/mkconfig.rb
+++ b/tool/mkconfig.rb
@@ -121,6 +121,7 @@
     val = val.gsub(/\$(?:\$|\{?(\w+)\}?)/) {$1 ? "$(#{$1})" : $&}.dump
     case name
     when /^prefix$/
+      v_runtime[:prefix] = val.slice(1..-2)
       val = "(TOPDIR || DESTDIR + #{val})"
     when /^ARCH_FLAG$/
       val = "arch_flag || #{val}" if universal
@@ -193,6 +194,7 @@
     x
   end
 end
+v_others.map!{|x| x.gsub(v_runtime[:prefix], "$(prefix)") }
 v_others.compact!
 
 if $install_name
