--- a/lib/rubygems/security.rb	(revision 53418)
+++ b/lib/rubygems/security.rb	(revision 53419)
@@ -340,7 +340,7 @@
   # Digest algorithm used to sign gems
 
   DIGEST_ALGORITHM =
-    if defined?(OpenSSL::Digest) then
+    if defined?(OpenSSL::Digest::SHA1) then
       OpenSSL::Digest::SHA1
     end
 
@@ -356,7 +356,7 @@
   # Algorithm for creating the key pair used to sign gems
 
   KEY_ALGORITHM =
-    if defined?(OpenSSL::PKey) then
+    if defined?(OpenSSL::PKey::RSA) then
       OpenSSL::PKey::RSA
     end
 
