diff --git a/lib/logger.rb b/lib/logger.rb
index 15d95fc..69f0d7b 100644
--- a/lib/logger.rb
+++ b/lib/logger.rb
@@ -171,7 +171,13 @@ require 'monitor'
 class Logger
   VERSION = "1.2.6"
   id, name, rev = %w$Id$
-  ProgName = "#{name.chomp(",v")}/#{rev}"
+  if name
+    name = name.chomp(",v")
+  else
+    name = File.basename(__FILE__)
+  end
+  rev ||= "v#{VERSION}"
+  ProgName = "#{name}/#{rev}"
 
   class Error < RuntimeError; end
   class ShiftingError < Error; end
