diff --git a/test/drb/drbtest.rb b/test/drb/drbtest.rb
index bc16ab1..c12b168 100644
--- a/test/drb/drbtest.rb
+++ b/test/drb/drbtest.rb
@@ -22,7 +22,7 @@ class DRbService
   %w(ut_drb.rb ut_array.rb ut_port.rb ut_large.rb ut_safe1.rb ut_eval.rb).each do |nm|
     add_service_command(nm)
   end
-  @server = @@server = DRb::DRbServer.new('druby://localhost:0', @@manager, {})
+  @server = @@server = DRb::DRbServer.new('druby://127.0.0.1:0', @@manager, {})
   @@manager.uri = @@server.uri
   def self.manager
     @@manager
@@ -79,16 +79,16 @@ module DRbCore
   end
 
   def test_00_DRbObject
-    ro = DRbObject.new(nil, 'druby://localhost:12345')
-    assert_equal('druby://localhost:12345', ro.__drburi)
+    ro = DRbObject.new(nil, 'druby://127.0.0.1:12345')
+    assert_equal('druby://127.0.0.1:12345', ro.__drburi)
     assert_equal(nil, ro.__drbref)
 
-    ro = DRbObject.new_with_uri('druby://localhost:12345')
-    assert_equal('druby://localhost:12345', ro.__drburi)
+    ro = DRbObject.new_with_uri('druby://127.0.0.1:12345')
+    assert_equal('druby://127.0.0.1:12345', ro.__drburi)
     assert_equal(nil, ro.__drbref)
 
-    ro = DRbObject.new_with_uri('druby://localhost:12345?foobar')
-    assert_equal('druby://localhost:12345', ro.__drburi)
+    ro = DRbObject.new_with_uri('druby://127.0.0.1:12345?foobar')
+    assert_equal('druby://127.0.0.1:12345', ro.__drburi)
     assert_equal(DRb::DRbURIOption.new('foobar'), ro.__drbref)
   end
 
diff --git a/test/drb/ut_drb.rb b/test/drb/ut_drb.rb
index f5720cf..265713d 100644
--- a/test/drb/ut_drb.rb
+++ b/test/drb/ut_drb.rb
@@ -154,7 +154,7 @@ if __FILE__ == $0
 
   DRb::DRbServer.default_argc_limit(8)
   DRb::DRbServer.default_load_limit(4096)
-  DRb.start_service('druby://localhost:0', DRbEx.new)
+  DRb.start_service('druby://127.0.0.1:0', DRbEx.new)
   es = DRb::ExtServ.new(ARGV.shift, ARGV.shift)
   DRb.thread.join
 end
diff --git a/test/drb/ut_drb_drbssl.rb b/test/drb/ut_drb_drbssl.rb
index 0a2191e..bca3012 100644
--- a/test/drb/ut_drb_drbssl.rb
+++ b/test/drb/ut_drb_drbssl.rb
@@ -18,7 +18,7 @@ if __FILE__ == $0
 
   DRb::DRbServer.default_argc_limit(8)
   DRb::DRbServer.default_load_limit(4096)
-  DRb.start_service('drbssl://localhost:0', DRbEx.new, config)
+  DRb.start_service('drbssl://127.0.0.1:0', DRbEx.new, config)
   es = DRb::ExtServ.new(ARGV.shift, ARGV.shift)
   DRb.thread.join
 end
diff --git a/test/drb/ut_eval.rb b/test/drb/ut_eval.rb
index 4df963e..9127939 100644
--- a/test/drb/ut_eval.rb
+++ b/test/drb/ut_eval.rb
@@ -3,7 +3,7 @@ require 'drb/extserv'
 
 class EvalAttack
   def initialize
-    @four = DRb::DRbServer.new('druby://localhost:0', self, {:safe_level => 4})
+    @four = DRb::DRbServer.new('druby://127.0.0.1:0', self, {:safe_level => 4})
   end
 
   def four
@@ -25,7 +25,7 @@ if __FILE__ == $0
 
   $SAFE = 1
 
-  DRb.start_service('druby://localhost:0', EvalAttack.new, {:safe_level => 2})
+  DRb.start_service('druby://127.0.0.1:0', EvalAttack.new, {:safe_level => 2})
   es = DRb::ExtServ.new(ARGV.shift, ARGV.shift)
   DRb.thread.join
 end
diff --git a/test/drb/ut_large.rb b/test/drb/ut_large.rb
index d6717c5..0aefd1b 100644
--- a/test/drb/ut_large.rb
+++ b/test/drb/ut_large.rb
@@ -31,7 +31,7 @@ if __FILE__ == $0
 
   DRb::DRbServer.default_argc_limit(3)
   DRb::DRbServer.default_load_limit(100000)
-  DRb.start_service('druby://localhost:0', DRbLarge.new)
+  DRb.start_service('druby://127.0.0.1:0', DRbLarge.new)
   es = DRb::ExtServ.new(ARGV.shift, ARGV.shift)
   DRb.thread.join
 end
diff --git a/test/drb/ut_safe1.rb b/test/drb/ut_safe1.rb
index 4df8e1e..ee097f6 100644
--- a/test/drb/ut_safe1.rb
+++ b/test/drb/ut_safe1.rb
@@ -8,7 +8,7 @@ if __FILE__ == $0
     it
   end
 
-  DRb.start_service('druby://localhost:0', [1, 2, 'III', 4, "five", 6], 
+  DRb.start_service('druby://127.0.0.1:0', [1, 2, 'III', 4, "five", 6], 
                     {:safe_level => 1})
   es = DRb::ExtServ.new(ARGV.shift, ARGV.shift)
   DRb.thread.join
diff --git a/test/net/pop/test_pop.rb b/test/net/pop/test_pop.rb
index c8aa9a8..129ca08 100644
--- a/test/net/pop/test_pop.rb
+++ b/test/net/pop/test_pop.rb
@@ -3,10 +3,14 @@ require 'test/unit'
 require 'digest/md5'
 
 class TestPOP < Test::Unit::TestCase
+  def localhost
+    '127.0.0.1'
+  end
+
   def setup
     @users = {'user' => 'pass' }
     @ok_user = 'user'
-    @stamp_base = "#{$$}.#{Time.now.to_i}@localhost"
+    @stamp_base = "#{$$}.#{Time.now.to_i}@#{localhost}"
   end
 
   def test_pop_auth_ok
@@ -64,7 +68,7 @@ class TestPOP < Test::Unit::TestCase
   end
 
   def pop_test(apop=false)
-    host = 'localhost'
+    host = localhost
     server = TCPServer.new(host, 0)
     port = server.addr[1]
     thread = Thread.start do
