--- a/tool/rbinstall.rb
+++ b/tool/rbinstall.rb
@@ -719,15 +719,15 @@ install?(:ext, :comm, :gem) do
     directories = Gem.ensure_gem_subdirectories(gem_dir, :mode => $dir_mode)
     prepare "bundle gems", gem_dir, directories
     Dir.glob(srcdir+'/gems/*.gem').each do |gem|
-      Gem.install gem, Gem::Requirement.default, :install_dir => with_destdir(Gem.dir), :domain => :local, :ignore_dependencies => true
+      Gem.install gem, Gem::Requirement.default, :install_dir => with_destdir(gem_dir), :domain => :local, :ignore_dependencies => true
       gemname = Pathname(gem).basename
       puts "#{" "*30}#{gemname}"
     end
     # fix directory permissions
     # TODO: Gem.install should accept :dir_mode option or something
-    File.chmod($dir_mode, *Dir.glob(with_destdir(Gem.dir)+"/**/"))
+    File.chmod($dir_mode, *Dir.glob(with_destdir(gem_dir)+"/**/"))
     # fix .gemspec permissions
-    File.chmod($data_mode, *Dir.glob(with_destdir(Gem.dir)+"/specifications/*.gemspec"))
+    File.chmod($data_mode, *Dir.glob(with_destdir(gem_dir)+"/specifications/*.gemspec"))
   else
     puts "skip installing bundle gems because of lacking zlib"
   end
