diff --git a/lib/mkmf.rb b/lib/mkmf.rb
index f4f84a6..f73f7ed 100644
--- a/lib/mkmf.rb
+++ b/lib/mkmf.rb
@@ -2191,7 +2191,7 @@ def init_mkmf(config = CONFIG, rbconfig = RbConfig::CONFIG)
   $LOCAL_LIBS = ""
 
   $cleanfiles = config_string('CLEANFILES') {|s| Shellwords.shellwords(s)} || []
-  $cleanfiles << "mkmf.log"
+  $cleanfiles << "mkmf.log .*.time"
   $distcleanfiles = config_string('DISTCLEANFILES') {|s| Shellwords.shellwords(s)} || []
   $distcleandirs = config_string('DISTCLEANDIRS') {|s| Shellwords.shellwords(s)} || []
 
