diff --git a/gc.c b/gc.c
index 8ab287e..a77ef2c 100644
--- a/gc.c
+++ b/gc.c
@@ -1029,6 +1029,7 @@ ruby_xfree(void *x)
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_enable(void)
 {
@@ -1051,6 +1052,7 @@ rb_gc_enable(void)
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_disable(void)
 {
@@ -1073,6 +1075,7 @@ rb_gc_disable(void)
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_enable_stats()
 {
@@ -1094,6 +1097,7 @@ rb_gc_enable_stats()
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_disable_stats()
 {
@@ -1113,6 +1117,7 @@ rb_gc_disable_stats()
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_stats_enabled()
 {
@@ -1132,6 +1137,7 @@ rb_gc_stats_enabled()
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_clear_stats()
 {
@@ -1199,6 +1205,7 @@ rb_gc_num_allocations()
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_enable_trace()
 {
@@ -1220,6 +1227,7 @@ rb_gc_enable_trace()
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_disable_trace()
 {
@@ -1239,6 +1247,7 @@ rb_gc_disable_trace()
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_trace_enabled()
 {
@@ -1267,6 +1276,7 @@ const char* GC_LOGFILE_IVAR = "@gc_logfile_name";
  *
  */

+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_log_file(int argc, VALUE *argv, VALUE self)
 {
