diff -Naurp ruby-1.8.7-p374.orig/ext/openssl/ossl.c ruby-1.8.7-p374/ext/openssl/ossl.c
--- ruby-1.8.7-p374.orig/ext/openssl/ossl.c	2022-08-16 11:48:06.331402171 -0700
+++ ruby-1.8.7-p374/ext/openssl/ossl.c	2022-08-16 11:50:55.835927272 -0700
@@ -391,18 +391,8 @@ ossl_debug_get(VALUE self)
 static VALUE
 ossl_debug_set(VALUE self, VALUE val)
 {
-    VALUE old = dOSSL;
-    dOSSL = val;
+    dOSSL = RTEST(val) ? Qtrue : Qfalse;
 	
-    if (old != dOSSL) {
-	if (dOSSL == Qtrue) {
-	    CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);
-	    fprintf(stderr, "OSSL_DEBUG: IS NOW ON!\n");
-	} else if (old == Qtrue) {
-	    CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_OFF);
-	    fprintf(stderr, "OSSL_DEBUG: IS NOW OFF!\n");
-	}
-    }
     return val;
 }
 
diff -Naurp ruby-1.8.7-p374.orig/ext/openssl/ossl_pkey_rsa.c ruby-1.8.7-p374/ext/openssl/ossl_pkey_rsa.c
--- ruby-1.8.7-p374.orig/ext/openssl/ossl_pkey_rsa.c	2022-08-16 11:48:06.331402171 -0700
+++ ruby-1.8.7-p374/ext/openssl/ossl_pkey_rsa.c	2022-08-16 11:49:41.881956217 -0700
@@ -636,7 +636,6 @@ Init_ossl_rsa()
     rb_define_method(cRSA, "params", ossl_rsa_get_params, 0);
 
     DefRSAConst(PKCS1_PADDING);
-    DefRSAConst(SSLV23_PADDING);
     DefRSAConst(NO_PADDING);
     DefRSAConst(PKCS1_OAEP_PADDING);
 
