Index: ext/readline/readline.c
===================================================================
--- ext/readline/readline.c     (revision 26664)
+++ ext/readline/readline.c     (working copy)
@@ -833,6 +833,12 @@
 #ifdef HAVE_RL_EVENT_HOOK
     rl_event_hook = readline_event;
 #endif
+#ifdef HAVE_RL_CATCH_SIGNALS
+    rl_catch_signals = 0;
+#endif
+#ifdef HAVE_RL_CATCH_SIGWINCH
+    rl_catch_sigwinch = 0;
+#endif
 #ifdef HAVE_RL_CLEAR_SIGNALS
     rl_clear_signals();
 #endif
Index: ext/readline/extconf.rb
===================================================================
--- ext/readline/extconf.rb     (revision 26664)
+++ ext/readline/extconf.rb     (working copy)
@@ -59,6 +59,9 @@
 have_readline_var("rl_attempted_completion_over")
 have_readline_var("rl_library_version")
 have_readline_var("rl_event_hook")
+# workaround for native windows.
+/mswin|bccwin|mingw/ !~ RUBY_PLATFORM && have_readline_var("rl_catch_sigwinch")
+/mswin|bccwin|mingw/ !~ RUBY_PLATFORM && have_readline_var("rl_catch_signals")
 have_readline_func("rl_cleanup_after_signal")
 have_readline_func("rl_clear_signals")
 have_readline_func("rl_vi_editing_mode")
