diff --git a/thread.c b/thread.c
index 4022f60..353d524 100644
--- a/thread.c
+++ b/thread.c
@@ -549,6 +549,7 @@ thread_create_core(VALUE thval, VALUE args, VALUE (*fn)(ANYARGS))
 
     th->priority = GET_THREAD()->priority;
     th->thgroup = GET_THREAD()->thgroup;
+    th->event_flags = GET_THREAD()->event_flags;
 
     native_mutex_initialize(&th->interrupt_lock);
     if (GET_VM()->event_hooks != NULL)
