--- a/installer.rb
+++ b/installer.rb
@@ -281,7 +281,7 @@
 		if install_autoconf_package('source', 'Ruby Enterprise Edition')
 			# Some installed files may have wrong permissions
 			# (not world-readable). So we fix this.
-			if sh("chmod -R g+r,o+r,o-w #{@destdir}#{@prefix}/lib/ruby")
+			if sh("chmod -R g+r,o+r,o-w #{@destdir}#{@prefix}/lib*/ruby")
 				if @using_system_allocator_library &&
 				   !sh("install source/libsystem_allocator.dylib #{@destdir}#{@prefix}/lib/")
 					return false
@@ -307,7 +307,7 @@
 	def install_rubygems
 		# We might be installing into a fakeroot, so add the fakeroot's library
 		# search paths to RUBYLIB so that gem installation will work.
-		basedir = "#{@destdir}#{@prefix}/lib/ruby"
+		basedir = "#{@destdir}#{@prefix}/lib*/ruby"
 		libdir = "#{basedir}/1.8"
 		archname = File.basename(File.dirname(Dir["#{libdir}/*/thread.#{PlatformInfo::RUBYLIBEXT}"].first))
 		extlibdir = "#{libdir}/#{archname}"
