Index: gem_prelude.rb
===================================================================
--- a/gem_prelude.rb	(revision 25926)
+++ b/gem_prelude.rb	(working copy)
@@ -67,7 +67,7 @@
 
     def self.set_home(home)
       home = home.gsub File::ALT_SEPARATOR, File::SEPARATOR if File::ALT_SEPARATOR
-      @gem_home = home.force_encoding(Encoding.find('filesystem'))
+      @gem_home = home.dup.force_encoding(Encoding.find('filesystem'))
     end
 
     def self.set_paths(gpaths)
