diff --git a/eval.c b/eval.c
index 284e486..00bcd36 100644
--- a/eval.c
+++ b/eval.c
@@ -12461,6 +12461,11 @@ rb_thread_start_0(fn, arg, th)
 	curr_thread->next = th;
 	th->priority = curr_thread->priority;
 	th->thgroup = curr_thread->thgroup;
+#if defined(HAVE_SETITIMER) || defined(_THREAD_SAFE)
+        if (!thread_init) {
+            rb_thread_start_timer();
+        }
+#endif
     }
     START_TIMER();
 
@@ -13188,7 +13193,9 @@ rb_thread_atfork()
     main_thread = curr_thread;
     curr_thread->next = curr_thread;
     curr_thread->prev = curr_thread;
-    STOP_TIMER();
+#if defined(HAVE_SETITIMER) || defined(_THREAD_SAFE)
+    rb_thread_stop_timer();
+#endif
 }
 
 
