diff -Naurp ruby-2.5.9.orig/ext/openssl/deprecation.rb ruby-2.5.9/ext/openssl/deprecation.rb
--- ruby-2.5.9.orig/ext/openssl/deprecation.rb	2021-04-05 04:46:35.000000000 -0700
+++ ruby-2.5.9/ext/openssl/deprecation.rb	2022-08-19 12:02:59.603496778 -0700
@@ -2,7 +2,7 @@
 module OpenSSL
   def self.deprecated_warning_flag
     unless flag = (@deprecated_warning_flag ||= nil)
-      if try_compile("", flag = "-Werror=deprecated-declarations")
+      if false and try_compile("", flag = "-Werror=deprecated-declarations")
         $warnflags << " #{flag}"
       else
         flag = ""
diff -Naurp ruby-2.5.9.orig/ext/openssl/ossl_pkey_rsa.c ruby-2.5.9/ext/openssl/ossl_pkey_rsa.c
--- ruby-2.5.9.orig/ext/openssl/ossl_pkey_rsa.c	2021-04-05 04:46:35.000000000 -0700
+++ ruby-2.5.9/ext/openssl/ossl_pkey_rsa.c	2022-08-19 12:02:33.394772672 -0700
@@ -939,7 +939,8 @@ Init_ossl_rsa(void)
     rb_define_method(cRSA, "params", ossl_rsa_get_params, 0);
 
     DefRSAConst(PKCS1_PADDING);
-    DefRSAConst(SSLV23_PADDING);
+    // Bravenet - Disable, to allow compilation with OpenSSL 3.0+.  Removed from OpenSSL 3.0+
+    //DefRSAConst(SSLV23_PADDING);
     DefRSAConst(NO_PADDING);
     DefRSAConst(PKCS1_OAEP_PADDING);
 
