--- trunk/configure 2013-06-27 13:16:17.000000000 +0200
+++ trunk/configure 2013-09-08 20:47:28.911451551 +0200
@@ -7424,7 +7424,7 @@
   unset ac_c_werror_flag
 fi

- if test "x$stack_protector" = xyes; then
+ if test "x$stack_protector" = xnotbroken; then

     save_LDFLAGS="$LDFLAGS"
     LDFLAGS="$LDFLAGS -fstack-protector"
@@ -7475,7 +7475,7 @@
  fi
      ;;
 esac
-    if test "x$stack_protector" = xyes; then
+    if test "x$stack_protector" = xnotbroken; then
  # RUBY_APPEND_OPTION(XCFLAGS, -fstack-protector)
  case " ${XCFLAGS-} " in #(
   *' -fstack-protector '*) :
