--- a/random.c
+++ b/random.c
@@ -521,7 +521,7 @@
     CryptGenRandom(prov, size, seed);
     return 0;
 }
-#elif defined __linux__ && defined SYS_getrandom
+#elif defined __linux__ && defined __NR_getrandom
 #include <linux/random.h>

 # ifndef GRND_NONBLOCK
@@ -538,7 +538,7 @@
 	if (!need_secure)
 	    flags = GRND_NONBLOCK;
 	errno = 0;
-	ret = syscall(SYS_getrandom, seed, size, flags);
+	ret = syscall(__NR_getrandom, seed, size, flags);
 	if (errno == ENOSYS) {
 	    ATOMIC_SET(try_syscall, 0);
 	    return -1;
