diff --git a/gc.c b/gc.c
index e526125..0a8ee3a 100644
--- a/gc.c
+++ b/gc.c
@@ -1161,6 +1161,7 @@ rb_gc_clear_stats()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_allocated_size()
 {
@@ -1182,6 +1183,7 @@ rb_gc_allocated_size()
  *     GC.num_allocations    #=> 150
  *
  */
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_num_allocations()
 {
@@ -1515,6 +1517,7 @@ rb_gc_heap_slots()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_collections()
 {
@@ -1533,6 +1536,7 @@ rb_gc_collections()
  *
  */
 
+RUBY_FUNC_EXPORTED
 VALUE
 rb_gc_time()
 {
@@ -4147,6 +4151,7 @@ VALUE os_live_objects(VALUE self)
     return ULONG2NUM(live_objects);
 }
 
+RUBY_FUNC_EXPORTED
 unsigned long rb_os_live_objects()
 {
     rb_objspace_t *objspace = &rb_objspace;
@@ -4171,6 +4176,7 @@ VALUE os_allocated_objects(VALUE self)
 #endif
 }
 
+RUBY_FUNC_EXPORTED
 unsigned LONG_LONG rb_os_allocated_objects()
 {
     rb_objspace_t *objspace = &rb_objspace;
