diff --git a/class.c b/class.c
index 63fdd76..f5cb4a1 100644
--- a/class.c
+++ b/class.c
@@ -855,6 +855,7 @@ class_instance_method_list(int argc, VALUE *argv, VALUE mod, int obj, int (*func
     }
     ary = rb_ary_new();
     sa_foreach(&list, func, ary);
+    sa_clear(&list);
 
     return ary;
 }
@@ -1089,6 +1090,7 @@ rb_obj_singleton_methods(int argc, VALUE *argv, VALUE obj)
     }
     ary = rb_ary_new();
     sa_foreach(&list, ins_methods_i, ary);
+    sa_clear(&list);
 
     return ary;
 }
