diff --git a/eval.c b/eval.c
index 9ceb833..fb5afc0 100644
--- a/eval.c
+++ b/eval.c
@@ -380,8 +380,7 @@ setup_exception(rb_thread_t *th, int tag, volatile VALUE mesg)
     if (file) line = rb_sourceline();
     if (file && !NIL_P(mesg)) {
 	if (mesg == sysstack_error) {
-	    at = rb_enc_sprintf(rb_usascii_encoding(), "%s:%d", file, line);
-	    at = rb_ary_new3(1, at);
+	    at = rb_make_backtrace();
 	    rb_iv_set(mesg, "bt", at);
 	}
 	else {
