commit 9fa6fc6ac323d68ab4166f27247a290e2877e774
Author: Stefan Kaes <skaes@railsexpress.de>
Date:   Sun Dec 11 06:57:31 2011 +0100

    trace bmethods

diff --git a/vm_insnhelper.c b/vm_insnhelper.c
index 6b78394..bc84327 100644
--- a/vm_insnhelper.c
+++ b/vm_insnhelper.c
@@ -422,8 +422,13 @@ vm_call_bmethod(rb_thread_t *th, VALUE recv, int argc, const VALUE *argv,
     /* control block frame */
     th->passed_me = me;
 
+    EXEC_EVENT_HOOK(th, RUBY_EVENT_CALL, recv, me->called_id, me->klass);
+
     GetProcPtr(me->def->body.proc, proc);
     val = rb_vm_invoke_proc(th, proc, recv, argc, argv, blockptr);
+
+    EXEC_EVENT_HOOK(th, RUBY_EVENT_RETURN, recv, me->called_id, me->klass);
+
     return val;
 }
 
