--- a/vendor/google-perftools-1.2/src/tcmalloc.cc       2011-02-24 02:09:12.000000000 -0800
+++ b/vendor/google-perftools-1.2/src/tcmalloc.cc       2011-11-11 12:46:13.704738476 -0800
@@ -1669,5 +1669,5 @@
   MallocHook::InvokeNewHook(result, size);
   return result;
 }
-void *(*__memalign_hook)(size_t, size_t, const void *) = MemalignOverride;
+void *(* volatile __memalign_hook)(size_t, size_t, const void *) = MemalignOverride;
 #endif  // #ifndef TCMALLOC_FOR_DEBUGALLOCATION
